LSP.sublime-settings

```diff
@@ -105,6 +105,13 @@
   // Valid values are "dot", "circle", "bookmark", "sign" or ""
   "diagnostics_gutter_marker": "dot",
 
+  // Enable semantic highlighting in addition to standard syntax highlighting (experimental!).
+  // Note: Must be supported by the language server and also requires a special rule
+  // in the color scheme to work. Please see the documentation under
+  //   https://lsp.sublimetext.io/customization/#semantic-highlighting
+  // for a description about how to configure your color scheme for semantic highlighting.
+  "semantic_highlighting": false,
```

Comment:
No. Third-party color schemes can already come with that special rule built-in. For example I have already added a similar rule for the documentHighlights into my own color scheme, so that users just need to set the "document_highlight_style" setting to "fill" and it will work out of the box with a nice style. I would do the same for the semantic highlighting then, if/when it gets merged. Users still need a way to disable it then. The way over "disabled_capabilities" is no real alternative, because it is per server, and it is too complicated for users who are not very familiar with the LSP. Also I would say it should be disabled by default, because with the two servers I used for testing, the disadvantages from semantic highlighting seem bigger to me than the advantages.

Reply:
Also this special color scheme rule is just a temporary requirement for the "add_regions" workaround, and it can be removed as soon as there is a dedicated ST API for syntax highlighting / foreground color adjustments which we can use instead (perhaps never ;) ).

Reply:
- Two steps is too confusing for end-users. Actually, one step is already confusing for most users I would think.
- We want less settings, not more settings. More settings means more things to check when refactoring, implying a larger maintenance burden, increasing reluctance to moving the project forward, decreasing velocity.
- There is overlap with disabled_capabilities. When troubleshooting for yet another user in the Discord, I would now have to ask "do you have semantic_highlighting set to false/true" in addition to "what is your disabled_capabilities"

I understand your logic, but your Brackets color scheme is the _only_ color scheme that would start handling semantic highlighting automatically. Perhaps that's a good thing? Users enjoy it when things Just Work (tm). Discoverability is increased this way. Maybe we should add more documentation about disabled_capabilities in the documentation?

Maybe I'm just too stubborn in this regard? I really do worry having even more settings decreases velocity.

> Also this special color scheme rule is just a temporary requirement for the "add_regions" workaround, and it can be removed as soon as there is a dedicated ST API for syntax highlighting / foreground color adjustments which we can use instead

I don't think we'll ever see such an API. Because semantic highlighting changes the meaning of tokens, the "Goto Symbol" and "Goto Symbol In Project" would have to become dynamic as well. At that point SublimeHQ might as well write a native LSP client. They're already reluctant to add a file watcher API.

Reply:
> * Two steps is too confusing for end-users. Actually, one step is already confusing for most users I would think.

That is why I already added the required rule to the built-in color schemes (it was reverted based on your last comment). Then users of the built-in color schemes would have to enable only the boolean setting.

I would also argue that it is much easier to find this setting in the LSP.sublime-settings file and read its description, than to know about that you need to add a special customization to your color scheme, as a requirenment to enable semantic highlighting. If we remove the boolean setting, the only way to find out about this, is if someone reads the "customization" section in our LSP documentation. And nobody reads documentation anyway. :trollface:

> * We want less settings, not more settings. More settings means more things to check when refactoring, implying a larger maintenance burden, increasing reluctance to moving the project forward, decreasing velocity.

I see that you are a bit reluctant to add new settings in genereal, and I have a different opinion on it. I prefer to have as many settings as possible (when it makes sense), so that users can configure a software/plugin/theme etc. to their liking. I don't really do much refactoring of this package, but I also don't think the settings are a lot of maintenance burden. If any, having fewer settings will lead to complaints or feature request to make things configurable, see for example https://github.com/sublimelsp/LSP/issues/1729 when the styling options for diagnostics were removed. That example ultimately led to a new quite big new contribution with the diagnostics manager and diagnostics quick panel, but I don't think that is what you had in mind here ;)

> * There is overlap with disabled_capabilities.

Yes there is overlap, but as you wrote yourself, the disabled_capabilities is too complicated and hidden for most end users. I doubt that more than a handful of users use it.

For example, why is "disabled_capabilities" a JSON object, and not a simple array? From being an object, I assume I can also set the value for a capability to `false`, maybe to enable a capability explicitly? But then why is it called "disabled_capabilities" and not just "capabilities" (with the opposite boolean value)?

And as I already wrote, "disabled_capabilities" also has no way to let us disable a capability by default (I think). I would highly recommend to have semantic highlighting disabled by default. I only tested briefly with these servers, but what I saw with clangd and rust-analyzer:
* clangd doesn't use any token modifiers at all. That means that all highlighting information from sublime-syntaxes about whether an identifier is a definition or usage is lost, meaning for example that `entity.name.function` will be highlighted the same as `variable.function` after enabling semantic highlighting. Same with classes/types etc.
* rust-analyzer emits tokens for basically every non-whitespace character with all kinds of custom tokens like "semicolon", "brace", "parenthesis", ... which is in my opinion a vast misunderstanding of the purpose of semantic highlighting. It seems also to use wrong/questionable tokens in some cases, for examples arguments in function definitions are just "variable" instead of "parameter", which means that the usual highlighting of function arguments from the "variable.parameter" syntax scope is lost and they become plain white/black in most color schemes.

In general the highlighting from semantic tokens feels a bit laggy because of the delay from the server communication.

> When troubleshooting for yet another user in the Discord, I would now have to ask "do you have semantic_highlighting set to false/true" in addition to "what is your disabled_capabilities"

Seriously? We will have the following choice:
1. troubleshooting for users who say "LSP doesn't support semantic highlighting" -> answer: set "semantic_highlighting" setting to true
2. troubleshooting for users who say "with LSP syntax highlighting suddenly became very laggy and super weird" -> answer: you must open the correct LSP-foobar settings file, or the common LSP settings file, depending whether you use a helper plugin or not, then create a JSON object under the "clients", or at the top level object, respectively, then add a key-value pair with the special capability name, which can be found if you do an extensive search in the LSP specs, and finally set its value to true :wink:

> I understand your logic, but your Brackets color scheme is the *only* color scheme that would start handling semantic highlighting automatically.

Yes that is probably true, but if we don't have a way to disable it by default, I would not add it to my color scheme, due to the disadvantages listed above.

> I don't think we'll ever see such an API. Because semantic highlighting changes the meaning of tokens, the "Goto Symbol" and "Goto Symbol In Project" would have to become dynamic as well.

Probably not in the immediate future, but I wouldn't completely rule out that it gets implemented at some point. Maybe when more language servers start to support semantic highlighting and there is increased demand to have it available in ST too. I think the main objection for such an API is that users might blame ST for bad highlighting or delays while typing, when in reality it would be caused by the plugin.

But if I understand it correctly, the symbol locations are / can already be dynamically computed on demand, when you open the "Goto Symbol" or "Goto Symbol In Project" panel. Based on https://github.com/sublimehq/sublime_text/issues/4763#issuecomment-899313572.

Also you implicate that a potential API would have to be connected with ST's symbol locations. But that doesn't have to be the case, you can have an API to control just the highlighting style (probably via scopes), but the symbol locations can still be evaluated exclusively from the syntax scopes. Indeed if you want to also control symbol locations from the API, then it might be better to introduce another dedicated API function for it. The `textDocument/documentSymbol` and `workspace/symbol` requests would be more suited rather than semantic tokens in that case.

---

FWIW, VSCode has a user setting "editor.semanticHighlighting.enabled" with possible values `true`, `false`, and `"configuredByTheme"`. And there is another boolean setting in the theme to enable/disable semantic highlighting.

Reply:
Alright, let's then keep the boolean and revert ea80e28.

Reply:
> Probably not in the immediate future, but I wouldn't completely rule out that it gets implemented at some point. 

I have my doubts :) You see, there's this general approach from SublimeHQ to the ST API that plugins should take care of their own UI presentation. This allows SublimeHQ to put their head in the sand with respect to various editor concepts, like "diagnostics", "local symbols", "global symbols", "formatting". Plugins can just invent the concepts on their own and use general output panels, quick panels, on_pre_save, etc, to make it work.

I have come to realize that this is a pretty bad API design. If Package A and Package B both have their concept of diagnostics, then those two sets of diagnostics do not collaborate together nicely for the end-user. Even worse, both A and B have wasted their time writing all sorts of UI code to present those diagnostics. Just to highlight how big of a mess this really is:

- Default/exec.py re-invents the wheel on how diagnostics should be presented
- the built-in spellchecker re-invents the wheel on how diagnostics should be presented
- SublimeLinter maintainers spending their free time on UI presentation again
- LSP maintainers spending their free time on UI presentation again

> The textDocument/documentSymbol and workspace/symbol requests would be more suited rather than semantic tokens in that case.

I agree. But it gets even worse. If those were to be API endpoints, then in addition to SublimeHQ requiring to acknowledge the concept of "local symbols" and "global symbols" in the ST API, the callbacks would also need to support promises/futures. The current promise implementation for on_query_completions is ad-hoc. This implementation does not convince me that the engineers at SublimeHQ actually understand promises in a general manner. It took a _very_ long time to finally get that async on_query_completions. IMO the underlying problem is that async functions are not first-class citizens in Sublime Text, and furthermore requests for going in this direction is met with confusion or silence: https://github.com/sublimehq/sublime_text/issues/3129. So in conclusion the lack of acknowledgment of editor concepts, and the lack of asynchronicity as a first-class citizen make me feel like we have a long road ahead.

Compare this with VSCode: it acknowledges concepts. It provides clear provider endpoints for concepts ("hover", "diagnostics", "formatting"). Plugins mesh well together to extend those concepts in a cooperative way. Async functions are first-class citizens.